### PR TITLE
vopr: fix false positive when getting to recovering_head after version upgrade

### PR DESCRIPTION
**Pull request by matklad**
We have special code to disable storage faults when restarting a replica after a crash in the vopr. We also have similar code when restarting replica during upgrades but it doesn't quiet handle all special cases (in particular, it doesn't repair the WAL). Extract this special code into a special function to make sure that more of the logic is shared!

In particular, in this seed we have a cluster of two, where one replica gets `.recovering_head` in the release upgrade path

Seed: ./zig/zig build -Drelease vopr -- 11029999072009206202

### CODE REVIEW

File: src/vopr.zig [hunk 1]
```diff
@@ -1104,13 +1085,9 @@ pub const Simulator = struct {
         ) catch unreachable;
 
         if (replica.status == .recovering_head) {
-            // Even with faults disabled, a replica that was syncing before it crashed
-            // (or just recently finished syncing before it crashed) may wind up in
-            // status=recovering_head.
-            assert(fault or
-                replica.op < replica.op_checkpoint() or
-                replica.log_view < replica.superblock.working.vsr_state.sync_view or
```
Comment (matklad) [inline src/vopr.zig:1112]: Another artifact of the old state sync (or rather, my initial attempt at fixing it)

**Comment by matklad** [comment]
Kinda unhappy about this fix to be honest! This gets the seed going, and I think it _is_ an improvement, but I think it's a matter of time until VOPR finds a seed that works-around our repair attempts here because it _is_ possible that wal is un-repairable.

I am wondering if I should be scrapping this whole "try not to fault" logic here, and instead do a "core_missing_head" thing?

**Comment by matklad** [comment]
Requesting review mostly to brainstorm the proper solution here!

**Comment by sentientwaffle** [comment]
:thinking: My main issue is that this extra repair code shouldn't run during the `replica_test`s. (Which for me hints that `Cluster` is the wrong place for it).

> I am wondering if I should be scrapping this whole "try not to fault" logic here, and instead do a "core_missing_head" thing?

Yeah, I'm starting to lean that way too. My only worry is that it could end up "hiding" a real liveness bug if it was too broad -- i.e. if over time it began to cover scenarios that we weren't actually aware of --  but that seems manageable by making the verification strict.

**Comment by matklad** [comment]
I sort of convinced myself that the right fix here is to go in the opposite direction, and to “just” get rid of recovering head. That is, use that idea that replicas would _explocitly_ send nacks for the entire journal in a DVC. So that a recovering head replica is just a normal replica that simply dosnt nack and prepares past the head.

I worry that with our fault disabling logic we already might be masking a real liveness issue, as vopr seems to eager with repairing replicas if there’s a chance of one more recovering head

**Comment by matklad** [comment]
Discussed this during walk&talk, decided to let it be for the moment and revisit once this bites us again!

**Pull request closed without merging**